### PR TITLE
fix: Fixes the single host lag reporting case

### PR DESCRIPTION
**Pull request by AlanConfluent**
### Description 
Currently, when a single host exists, lag doesn't get reported.  This ensures lag always gets reported.  Also changes the default to not filter if there is no lag data.

### Testing done 
_Describe the testing strategy. Unit and integration tests are expected for any behavior changes._

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatAgent.java [hunk 1]
```diff
@@ -223,6 +223,9 @@ private void processHeartbeats(final long windowStart, final long windowEnd) {
           }
           return status;
         });
+        for (HostStatusListener listener : hostStatusListeners) {
```
Comment (vinothchandar) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatAgent.java]: is this related to the single host lag reporting case?

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/MaximumLagFilter.java [hunk 2]
```diff
@@ -66,8 +66,8 @@ public boolean filter(final KsqlHostInfo hostInfo) {
           final long offsetLag = Math.max(endOffset - hostLag.getCurrentOffsetPosition(), 0);
           return offsetLag <= allowedOffsetLag;
         })
-        // If we don't have lag info, we'll be conservative and not include the host
-        .orElse(false);
+        // If we don't have lag info, we'll be conservative and include the host
```
Comment (vinothchandar) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/MaximumLagFilter.java]: I think we should do this and then prime the cache on the remote host on-demand as we discussed..  This will ensure correct filtering in all cases.. 

Choosing to ignore nodes if we don't have lag , would mean that, on every bounce of the ksql server, queries will fail for few seconds until lags are propogated across the cluster again (bounced server will begin with an empty map).. Thus a rolling bounce of the cluster will see large error spikes.. This is a bad move IMO

Comment (vinothchandar) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/MaximumLagFilter.java]: Please file an issue for the the follow up work of priming the cache on the remote host, on demand...

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatAgent.java (hunk 1 above)
Comment (vinothchandar) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatAgent.java]: So previously, the heart beat agent was not notifying the lag agent?

Comment (AlanConfluent) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatAgent.java]: Yes, if there are no heartbeats received, it will still update the status of known hosts (localhost) and that needs to be reported to the listeners.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/MaximumLagFilter.java (hunk 2 above)
Comment (AlanConfluent) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/MaximumLagFilter.java]: Yeah, I agree with that.  I guess the only downside is that if we don't know the lag and we forward the request to that host, we may not be honoring the max lag configuration and may be returning a really stale value.

Comment (vinothchandar) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/MaximumLagFilter.java]: What I meant was, the remote host (i.e the one whose lag we dont know) can do the lag enforcement and fail the query and send an error response.. This lag exchange control plane is an optimization for not doing this round trip when we know the host is backed up/lagging significantly

Comment (AlanConfluent) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/MaximumLagFilter.java]: #4471 #4472 - Those are the two followup issues.

I see.  That makes sense.  I agree that it seems necessary to ensure the filtering if the client, in some way, is relying on that for some notion of consistency, etc.  If we're going to enforce this on the lagging host and the control plane is an optimization, then this makes sense.

**Comment by vpapavas** [comment]
Thank you @AlanConfluent! Does it make sense to add a functional test with a single server that catches this corner-case (and maybe others we have not found yet?)

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatAgent.java [hunk 3]
```diff
@@ -223,6 +223,9 @@ private void processHeartbeats(final long windowStart, final long windowEnd) {
...
+          listener.onHostStatusUpdated(getHostsStatus());
```
Comment (vinothchandar) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatAgent.java]: any exception handling needed here to ensure we don't stop notifying if one of  the callbacks throw an error?

Comment (AlanConfluent) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatAgent.java]: Done.  Factored out notification to a method where I now catch and log any errors.

**Comment by AlanConfluent** [comment]
> Changes in this PR itself sg to me. +1 to @vpapavas on adding more cases to functional test. Wondering if that can happen in another PR though..
> 
> Also could you share any local testing you may have done around this?

I have a functional test I'm working on.  Won't take long, but can make it another PR so we're not blocked on it.

I tested by modifying one of my functional tests to have one host and verified that it reported lag.  I also built the server and ran it locally with a single host and then looked at the clusterStatus as I produced messages to a topic and saw that the lag increased.
